### ISSUE TITLE
Image release pipeline failing unusually

### ISSUE THREAD

**Issue by robe070** [robe070/cookbooks#99]
`Release-44` is testing `w16d14-2-5` & `w19d14-2-5`. Build variables set appropriately. Maybe unnecessary as there are only artefacts for these 2 images. I was attempting to have the minimum jobs running. It seems there is no difference from having `Build ALL= True`. (Side issue. Please advise on best usage)

The main issue. `w16d14-2-5` fails to build very quickly with a VM provisioning error. I wonder if this is a pipeline resourcing issue as I have started 4 release pipelines for the 4 build runs that were required to build the 6 images.

**Or, more likely, its related to having failed builds still existing, possibly with different VM images (`44` used version 4, and `41` used 2. I know that when a VMSS changes the VM image the VMSS needs to be deleted and re-created. And also explain the speed with which it was at the stage of creating a VM in the VMSS, because all the other resources already existed.**

Here is the symptom:
![image](https://user-images.githubusercontent.com/4089160/100812603-8323dd00-3491-11eb-8706-24625dee63d3.png)

Also another of the 4 releases, `Release-47`,  failed differently, and at a similar time. Potentially related? Note that `Release-43`, which was building the same CI Run #464, had all jobs cancelled but the Release was not abandoned. The other 3 releases had the prior runs abandoned. ANd on further thought `Release-43` was started less than 30 mins previously to the start of `Release-47`, and the deployment actually completed at 11:32:57 am. `Release-47` was started at 11:19 am.

![image](https://user-images.githubusercontent.com/4089160/100812955-56bc9080-3492-11eb-89d5-0f935e33dd93.png)

So, if it is to do with the resource groups still existing, then there should be a check at the beginning to delete them if they already exist.

Holding off just deleting the RGs for your perusal. Need to have this investigated ASAP so the images may be published. Solution does not need to be implemented immediately if its not a show stopper, e.g. RG can be manually deleted. Just investigation is important.

Update:
I've updated the text with my thoughts on the causes and I seem to have come to a clear conclusion. So I'm going to manually delete the RGs and retry the releases. I'd still like your feedback, but not urgent. Like does it make sense to you?

**Comment by VarunChopraAccolite** [comment]
Hi @robe070 
The resource groups are not versioned to run in parallel for the release pipeline. It won't delete the resource group in case of failure to allow the user to RCA the failure using the resources(that's a manual step applicable for failed scenarios).

**Note**: In the case of build pipelines, it's versioned by appending the build number on the resource group.

Regards
Varun Chopra

**Comment by robe070** [comment]
Yes, I know why its left behind on failure. The question is why is it not deleted at the beginning of the next run?

**Comment by ravicharan07** [comment]
Hi @robe070 
 - As per discussion in yesterday call that in Delete resource group task we can see Custom condition set to only succeeded RG will delete not the Failed or canceled. 

- I am updating it with always() condition, which fulfill all success, failure and cancel conditions.
```
- Only when all previous dependencies have succeeded.

" and(succeeded(), eq(variables['Gate.IsEnabled'], 'True'), or(eq(variables['Build-w12r2d-14-2'], 'True'), eq(variables['Build-all'], 'True'))) "

- Even if a previous dependency has failed, unless the run was canceled. Use succeededOrFailed() in the YAML for this condition.

" and(succeededOrFailed(), eq(variables['Gate.IsEnabled'], 'True'), or(eq(variables['Build-w12r2d-14-2'], 'True'), eq(variables['Build-all'], 'True'))) "


- Even if a previous dependency has failed, even if the run was canceled. Use always() in the YAML for this condition.

" and(always(), eq(variables['Gate.IsEnabled'], 'True'), or(eq(variables['Build-w12r2d-14-2'], 'True'), eq(variables['Build-all'], 'True'))) "
```
Cc: @VarunChopraAccolite

Any suggestions, much appreciated.

**Comment by robe070** [comment]
That sounds incorrect.

Your task was not to change the task that was already there. Its to COPY the existing task to BEFORE the template deployment to ensure it is deleted when a previous run has failed. And the condition would be set to `always()`. And it should more often than not fail because the previous run succeeded, so it MUST continue on failure. And in fact it is best if it is not even recorded as a failed task, because non-existence is expected and is OK.

**Comment by VarunChopraAccolite** [comment]
hi @ravicharan07 
go for the below approach, make sure you perform a dummy run to verify your changes and share SS here.
```
- Only when all previous dependencies have succeeded.

" and(succeeded(), eq(variables['Gate.IsEnabled'], 'True'), or(eq(variables['Build-w12r2d-14-2'], 'True'), eq(variables['Build-all'], 'True'))) "
```

**Comment by ravicharan07** [comment]
Hi @robe070 , 

Added the new task(Delete RG with condition set to always()) in the image release pipeline. Please find the screen shot below. 
Kindly let me know if any changes required. 

![image](https://user-images.githubusercontent.com/72967378/102205103-b7b38200-3ef0-11eb-9a8c-69384b42d182.png)

**Comment by robe070** [comment]
Looks good
